quantum/os_detection.c

```diff
@@ -47,38 +47,47 @@ struct setups_data_t setups_data = {
 
 os_variant_t detected_os = OS_UNSURE;
 
+static bool detection_pending = true;
+
+void process_detected_host_os(os_variant_t os) {
+    process_detected_host_os_kb(os);
+}
+
+__attribute__((weak)) bool process_detected_host_os_kb(os_variant_t os) {
+    return process_detected_host_os_user(os);
+}
+
+__attribute__((weak)) bool process_detected_host_os_user(os_variant_t os) {
+    return true;
+}
+
 // Some collected sequences of wLength can be found in tests.
 void make_guess(void) {
-    if (setups_data.count < 3) {
-        return;
-    }
-    if (setups_data.cnt_ff >= 2 && setups_data.cnt_04 >= 1) {
-        detected_os = OS_WINDOWS;
-        return;
-    }
-    if (setups_data.count == setups_data.cnt_ff) {
-        // Linux has 3 packets with 0xFF.
-        detected_os = OS_LINUX;
-        return;
-    }
-    if (setups_data.count == 5 && setups_data.last_wlength == 0xFF && setups_data.cnt_ff == 1 && setups_data.cnt_02 == 2) {
-        detected_os = OS_MACOS;
-        return;
-    }
-    if (setups_data.count == 4 && setups_data.cnt_ff == 0 && setups_data.cnt_02 == 2) {
-        // iOS and iPadOS don't have the last 0xFF packet.
-        detected_os = OS_IOS;
-        return;
-    }
-    if (setups_data.cnt_ff == 0 && setups_data.cnt_02 == 3 && setups_data.cnt_04 == 1) {
-        // This is actually PS5.
-        detected_os = OS_LINUX;
-        return;
+    os_variant_t guessed = detected_os;
+    if (setups_data.count >= 3) {
+        if (setups_data.cnt_ff >= 2 && setups_data.cnt_04 >= 1) {
+            guessed = OS_WINDOWS;
+        } else if (setups_data.count == setups_data.cnt_ff) {
+            // Linux has 3 packets with 0xFF.
+            guessed = OS_LINUX;
+        } else if (setups_data.count == 5 && setups_data.last_wlength == 0xFF && setups_data.cnt_ff == 1 && setups_data.cnt_02 == 2) {
+            guessed = OS_MACOS;
+        } else if (setups_data.count == 4 && setups_data.cnt_ff == 0 && setups_data.cnt_02 == 2) {
+            // iOS and iPadOS don't have the last 0xFF packet.
+            guessed = OS_IOS;
+        } else if (setups_data.cnt_ff == 0 && setups_data.cnt_02 == 3 && setups_data.cnt_04 == 1) {
+            // This is actually PS5.
+            guessed = OS_LINUX;
+        } else if (setups_data.cnt_ff >= 1 && setups_data.cnt_02 == 0 && setups_data.cnt_04 == 0) {
+            // This is actually Quest 2 or Nintendo Switch.
+            guessed = OS_LINUX;
+        }
     }
-    if (setups_data.cnt_ff >= 1 && setups_data.cnt_02 == 0 && setups_data.cnt_04 == 0) {
-        // This is actually Quest 2 or Nintendo Switch.
-        detected_os = OS_LINUX;
-        return;
+    // report even OS_UNSURE if detection is running for the first time
+    if (detection_pending || guessed != detected_os) {
+        detection_pending = false;
+        process_detected_host_os(guessed);
```

Comment:
Invoking the callback from `make_guess()` won't work properly, because that code is called from the `get_usb_descriptor` callback, which gets called from the USB interrupt handler; it is not safe to call most QMK functions in that state.  (In fact, to be 100% safe, the `detected_os` variable should also be marked as `volatile`, because its value may be changed at any time when the USB interrupt comes.)

Instead of invoking the callback from `make_guess()`, you would probably need to add a task or deferred exec to check the `detected_os` variable for changes and invoke the callback from the main QMK loop.  E.g., see this:
https://github.com/qmk/qmk_firmware/blob/000ee31e2c64ad374b93be5a98f5529c48446188/users/drashna/drashna.c#L213-L263
https://github.com/qmk/qmk_firmware/blob/000ee31e2c64ad374b93be5a98f5529c48446188/users/drashna/callbacks.c#L65-L67

Reply:
Got it. I'll figure out a better place to put that.

I also need to figure out where such a thing could be called from whatever happens on the USB side when we switch machines using a KVM. I think it'd just be easier to soft reset the board when such an event is detected, though I'm not sure how to debug that to find out what is the flow of USB events when the KVM switches (seems to be too early on the keyboard initialization for me to get any console output or anything, of course).

Any idea on what to do on that front too?

Reply:
I've updated the code so the first callback is deferred until just before post-init, but I think I still need the deferred. Is that enough? Or should I still never call it from here because the state of the USB configuration, among other things, will likely still be changing?

Reply:
Ok, so the other task + a boolean to indicate when the detection runs might suffice then. 

Although 100ms is *likely* enough, I don't like relying on time so blindly. I'd rather have a marker to indicate when it runs and have the task be executed on the main loop if it is changed.

I'll change it later. I need to make sure the check whether to report or not is very, very cheap, because it'll be called quite a bit.

Reply:
> I also need to figure out where such a thing could be called from whatever happens on the USB side when we switch machines using a KVM.

Detecting that case might heavily depend on the particular KVM implementation (for some cases it might not be possible at all, if the KVM itself connects to the keyboard (so that it can intercept the key combinations for switching ports) and provides only an emulated keyboard to the hosts).

If the KVM treats the keyboard just as a generic USB device, you would probably see USB suspend first, then USB reset and the whole enumeration sequence some time later; in that case the sequence is hopefully generated by the actual computer, so you would be able to extract some information from it.

Reply:
On mine, it indeed is sent by the computer and the keyboard is just a USB device, but I can't get a suspend or anything of that sort. Sometimes it does reboot the keyboard, sometimes it does not, but it always recognizes the OS when I unplug the keyboard and plug it back in, so that part is working fine.

I'll investigate further at some point. For now, I'll only make this a bit nicer instead of relying on a deferred invocation.

Reply:
It might also be an USB reset without a suspend before it.

Apparently QMK may call `notify_usb_device_state_change_user(USB_DEVICE_STATE_INIT)` when USB reset happens, but that call is made from the main loop, at which time the device might have been already reconfigured.  Probably the OS detection code should be extended to handle the USB reset notification somehow (because it might be due to enumeration by a completely different host).

Reply:
Ok, will test it when I have the chance.

Worst case scenario, if I can trigger a soft reset that would be enough. Hopefully I can make it nice and without a reset.

But I think the enumeration (or at least the trigger to create a new USBDescriptor) isn't happening because the code as I have now would call the callback function whenever the detected OS changes and it doesn't trigger it on my keyboard when the KVM switches.

Reply:
I was curious about the timing, and changed some of the code (and change the deferred exec to run every 10ms), and the amount of time that it takes to settle is anywhere between 150ms to 500ms.  (this was calculated by running timer_read() and storing the value once it had a non-unsure response).  

Also, for the notify usb change state thing, I can confirm that it triggers on device change on my kvm, which I was also having the issue with detection.

Reply:
Ok, for the timing, I got back to you on [this comment](https://github.com/qmk/qmk_firmware/pull/21777#discussion_r1303656523). I ended up mixing the threads up.

As for the USB KVM thingy, I see my keyboard goes to `USB_DEVICE_STATE_INIT`, sometimes repeatedly, and then it finally moves to `USB_DEVICE_STATE_CONFIGURED` again.

And by that point, it seems the OS detection is always stable.
So, the solution? Well, I added some debounce to USB states too 😂 technically it's all a single debounce marker. It'll wait (by default 200ms) until there's no change to USB state, it is `USB_DEVICE_STATE_CONFIGURED` and then it'll report the OS detection.

As a bonus, I found one way of forcing the redetection, which is to soft reset the board when it gets a `USB_DEVICE_STATE_INIT` _after_ it once had been stable at `USB_DEVICE_STATE_CONFIGURED`.

So far, it's been working perfectly. Sometimes I might see some weirdness (LEDs, etc.) if I switch back and forth purposefully trying to break it, but it never soft-locks, at least.

Could you check if it works on your board too?

Reply:
just a heads up, `reset_keyboard()` jumps to bootloader, `soft_reset_keyboard()` is a soft reset, and returns to firmware.

Reply:
> just a heads up, `reset_keyboard()` jumps to bootloader, `soft_reset_keyboard()` is a soft reset, and returns to firmware.

@drashna yes, I realized that after a while and after realizing what `QK_BOOT` corresponds to. I removed `reset_keyboard()` for obvious reasons. If you set `OS_DETECTION_KEYBOARD_RESET` it'll call `soft_reset_keyboard()`. At least for my board, it fully solves the problem when switching devices with the KVM switch, since the soft reset resets the USB stuff too.

Could you please test that on your board and see if it triggers the reboot and that the debouncing is working correctly?

Reply:
@drashna oh, wait. Maybe we don't even need to reset. Let me try cleaning the struct where we store the wLength values first.

Reply:
@drashna never mind, it really needs to be a soft reset (or some other mechanism to reconfigure the device completely which I haven't found yet).

Reply:
> @drashna never mind, it really needs to be a soft reset (or some other mechanism to reconfigure the device completely which I haven't found yet).

Soft reset is probably best here, as anything else likely would require messing with the USB endpoint/driver, code